lib/vector-buffers/src/topology/channel/limited_queue.rs

```diff
@@ -54,9 +55,74 @@ impl<T> fmt::Display for TrySendError<T> {
 
 impl<T: fmt::Debug> std::error::Error for TrySendError<T> {}
 
+// Trait over common queue operations so implementation can be chosen at initialization phase
+trait QueueImpl<T>: Send + Sync + fmt::Debug {
+    fn push(&self, item: T);
+    fn pop(&self) -> Option<T>;
+}
+
+impl<T> QueueImpl<T> for ArrayQueue<T>
+where
+    T: Send + Sync + fmt::Debug,
+{
+    fn push(&self, item: T) {
+        self.push(item)
+            .unwrap_or_else(|_| unreachable!("acquired permits but channel reported being full."));
+    }
+
+    fn pop(&self) -> Option<T> {
+        self.pop()
+    }
+}
+
+impl<T> QueueImpl<T> for SegQueue<T>
+where
+    T: Send + Sync + fmt::Debug,
+{
+    fn push(&self, item: T) {
+        self.push(item);
+    }
+
+    fn pop(&self) -> Option<T> {
+        self.pop()
+    }
+}
+
+#[derive(Debug)]
+struct SizeTerms<T> {
+    method: fn(&T) -> usize,
+}
+
+impl<T> Clone for SizeTerms<T> {
+    fn clone(&self) -> Self {
+        Self {
+            method: self.method,
+        }
+    }
+}
+
+impl<T: InMemoryBufferable> SizeTerms<T> {
+    fn apply(&self, item: &T) -> usize {
+        (self.method)(item)
```

Comment:
Function pointers are actually nasty for performance, worse than conditional branches. In this case, it's also worse for clarity too. I think this could just be an enum, possibly even reusing `MemoryBufferSize` and adding a method.

Reply:
I didn't use an enum because then there is a conditional where I have to extract the value from the variant. In this way the logic to execute is fixed ahead of time, therefore a conditional is avoided.

Reply:
Did some research on this, seems like extracting from a variant will be still better off for performance.  For variant checks the complier can often inline the calls and branch predictions can be easier to perform for the CPU. Will modify.